server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationState.java

```diff
@@ -0,0 +1,485 @@
+        }
+        if (clusterState.version() <= lastPublishedVersion) {
+            logger.debug("handleClientValue: ignored request due to version mismatch " +
+                    "(expected: [term {} version >{}], actual: [term {} version {}])",
+                getCurrentTerm(), lastPublishedVersion, clusterState.term(), clusterState.version());
+            throw new CoordinationStateRejectedException("incoming cluster state version " + clusterState.version() +
+                " lower or equal to last published version " + lastPublishedVersion);
+        }
+
+        if (clusterState.getLastAcceptedConfiguration().equals(getLastAcceptedConfiguration()) == false
+            && getLastCommittedConfiguration().equals(getLastAcceptedConfiguration()) == false) {
+            logger.debug("handleClientValue: only allow reconfiguration while not already reconfiguring");
+            throw new CoordinationStateRejectedException("only allow reconfiguration while not already reconfiguring");
+        }
+        if (joinVotesHaveQuorumFor(clusterState.getLastAcceptedConfiguration()) == false) {
+            logger.debug("handleClientValue: only allow reconfiguration if joinVotes have quorum for new config");
+            throw new CoordinationStateRejectedException("only allow reconfiguration if joinVotes have quorum for new config");
+        }
+
+        assert clusterState.getLastCommittedConfiguration().equals(getLastCommittedConfiguration()) :
+            "last committed configuration should not change";
+
+        lastPublishedVersion = clusterState.version();
+        lastPublishedConfiguration = clusterState.getLastAcceptedConfiguration();
+        publishVotes = new VoteCollection();
+
+        logger.trace("handleClientValue: processing request for version [{}] and term [{}]", lastPublishedVersion, getCurrentTerm());
+
+        return new PublishRequest(clusterState);
+    }
+
+    /**
+     * May be called on receipt of a PublishRequest.
+     *
+     * @param publishRequest The publish request received.
+     * @return A PublishResponse which can be sent back to the sender of the PublishRequest.
+     * @throws CoordinationStateRejectedException if the arguments were incompatible with the current state of this object.
+     */
+    public PublishResponse handlePublishRequest(PublishRequest publishRequest) {
+        final ClusterState clusterState = publishRequest.getAcceptedState();
+        if (clusterState.term() != getCurrentTerm()) {
+            logger.debug("handlePublishRequest: ignored publish request due to term mismatch (expected: [{}], actual: [{}])",
+                getCurrentTerm(), clusterState.term());
+            throw new CoordinationStateRejectedException("incoming term " + clusterState.term() + " does not match current term " +
+                getCurrentTerm());
+        }
+        if (clusterState.term() == getLastAcceptedTerm() && clusterState.version() <= getLastAcceptedVersion()) {
```

Comment:
Tests pass with:
- `if (clusterState.version() <= getLastAcceptedVersion()) {`
- `if (clusterState.version() < getLastAcceptedVersion()) {`
- `if (clusterState.term() == getLastAcceptedTerm()) {`

Reply:
fixed in 0d84428d208